anatomy_of_variables/Anatomy of Variables.md

```diff
@@ -0,0 +1,229 @@
+
+
+We will start from what variables are in programming generally, then sight some examples in some programming languages and finish off on **Ruby Variables**
+
+## What are Variables?
+**Variables**, as defined by [wikipedia](https://en.wikipedia.org/wiki/Variable_(computer_science)) " are storage locations paired with an associated symbolic name, ***an identifier***, which contains some known or unknown quantity of information referred to as a value." In other words, a variable can be referred to as a ***container*** that holds values or information. The variable's name represents what information the variable contains. They are called variables because the represented information can change but the operations on the variable remain the same.
+
+Variables can be likened to Buckets where information can be kept while on the outside of the buckets are names. When referring to one of the buckets, we use the name of the bucket, not the data stored in the bucket.
+
+Variables are the same across all computer programming languages. What differentiates them is the way they are used and how they are been assigned.
+
+##Types of Variables
+There are two general types of variables. They are:
+
+* Local Variables
+* Global Variables
+
+**Local Variables**
+
+These are variables that are assigned within a function or a method. They can be used only within the method or function in which they have been assigned.
+
+```ruby
+def counter
+    count = 0
+    while count <= 10
+        puts count
+        count += 1
+    end
+end
+
+count #=> 0 1 2 3 4 5 6 7 8 9 10
+```
+
+This is a ruby method named `counter`. A local variable `count` was assigned in this method and it was used to print `0..10`. However if you call count outside this method, it would throw a `method or variable not defined` error.
+
+Here is the JavaScript version of the above code:
+
+```JavaScript
+function counter(){
+    var count = 0;
+    while (count <= 10) {
+       console.log(count);
+        count ++;
+    }
+}
+
+counter(); //=> 0 1 2 3 4 5 6 7 8 9 10
+```
+
+It is important to note that arguments passed into a method or function, are also treated as local variables. For example, the code snippet below works exactly as the one above.
+
+```ruby
+def counter(count)
+    while count <= 10
+        puts count
+        count += 1
+    end
+end
+
+counter(0); #=>0 1 2 3 4 5 6 7 8 9 10
+```
+
+Here is the JavaScript version of the code above.
+
+```JavaScript
+function counter(count){
+    while (count <= 10) {
+       console.log(count);
+        count ++;
+    }
+}
+
+var count = 0;
+counter(count); //=>0 1 2 3 4 5 6 7 8 9 10
+```
+
+**Global Variables**
+
+These are variables that can be used by two or more methods or function. These variables are assigned outside the methods or functions and they can be manipulated within methods or functions.
+
+The example below shows how global variables are assigned and used in Ruby and JavaScript.
+
+```ruby
+@count = 0
```

Comment:
Global variables in ruby starts with $. So we can have `$count`. @count is a different form of global variables called `instance variables`.